platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/mappingToJvm/DeclarativeReflectionToObjectConverter.kt

```diff
@@ -54,40 +56,44 @@ class DeclarativeReflectionToObjectConverter(
     val reflectionIdentityObjects = mutableMapOf<ObjectAccessKey, Any?>()
 
     private
-    fun objectByIdentity(key: ObjectAccessKey, newObject: () -> Any?): Any? {
+    val reflectionIdentityPublicTypes = mutableMapOf<ObjectAccessKey, KClass<*>?>()
+
+    private
+    fun objectByIdentity(key: ObjectAccessKey, newInstanceAndPublicType: () -> InstanceAndPublicType): InstanceAndPublicType {
         // This code does not use `computeIfAbsent` because `newObject` can make reentrant calls, leading to CME.
         // Also, it does not check for the value being null, because null values are potentially allowed
         if (key !in reflectionIdentityObjects) {
-            val newInstance = newObject()
-            reflectionIdentityObjects[key] = newInstance
-            return newInstance
+            val instanceAndPublicType = newInstanceAndPublicType()
+            reflectionIdentityObjects[key] = instanceAndPublicType.instance
+            reflectionIdentityPublicTypes[key] = instanceAndPublicType.publicType
+            return instanceAndPublicType
         } else {
-            return reflectionIdentityObjects[key]
+            return InstanceAndPublicType.of(reflectionIdentityObjects[key],  reflectionIdentityPublicTypes[key])
         }
     }
 
     private
     fun applyPropertyValue(receiver: ObjectOrigin, property: DataProperty, assigned: PropertyValueReflection) {
         when (assigned.assignmentMethod) {
-            is AssignmentMethod.Property -> setPropertyValue(receiver, property, getObjectByResolvedOrigin(assigned.value.objectOrigin))
+            is AssignmentMethod.Property -> setPropertyValue(receiver, property, getObjectByResolvedOrigin(assigned.value.objectOrigin).instance)
             is AssignmentMethod.BuilderFunction -> invokeBuilderFunction(receiver, assigned.assignmentMethod.function, assigned.value.objectOrigin)
             is AssignmentMethod.AsConstructed -> Unit // the value should have already been passed to the constructor or the factory function
         }
     }
 
     private
-    fun getObjectByResolvedOrigin(objectOrigin: ObjectOrigin): Any? {
+    fun getObjectByResolvedOrigin(objectOrigin: ObjectOrigin): InstanceAndPublicType {
         return when (objectOrigin) {
             is ObjectOrigin.DelegatingObjectOrigin -> getObjectByResolvedOrigin(objectOrigin.delegate)
-            is ObjectOrigin.ConstantOrigin -> objectOrigin.literal.value
-            is ObjectOrigin.EnumConstantOrigin -> getEnumConstant(objectOrigin)
-            is ObjectOrigin.External -> externalObjectsMap[objectOrigin.key] ?: error("no external object provided for external object key of ${objectOrigin.key}")
+            is ObjectOrigin.ConstantOrigin -> InstanceAndPublicType.of(objectOrigin.literal.value, objectOrigin.literal.type.constantType.kotlin)
+            is ObjectOrigin.EnumConstantOrigin -> InstanceAndPublicType.of(getEnumConstant(objectOrigin), Class.forName(objectOrigin.javaTypeName).kotlin)
```

Comment:
Using `Class.forName` without an explicit class loader might fail if the class we want to load by name comes from a child class loader (relatively to the class loader of the calling class). We should only load classes by name using an explicit class loader reference to the class loader obtained from the right class loader scope.

Reply:
Fixed.

Reply:
Incorporated!